include/dynamatic/Dialect/Handshake/HandshakeTypes.td

```diff
@@ -206,4 +206,172 @@ def ChannelOrSimpleControl : TypeConstraint<
+
+      // Here the front operand is of ChannelType.
+
+      // Ensure that data types of all operands are the same
+      Type frontDataType = mlir::cast<ChannelType>(types.front()).getDataType();
+      for (size_t i = 1; i < numTypes; i++) {
+        if (mlir::cast<ChannelType>(types[i]).getDataType() != frontDataType)
+          return false;
+      }
+      return true;
+    })
+  }] # "($" # name # ")">
+>;
+
+/// Multi-entity constraint ensuring that the front operand of variadic and non-variadic
+/// operand have the same data types.
+/// Also applies to ControlTypes, as some operations treat them interchangeably.
+class DataTypesMatchWithVariadicFront<string variadic, string nonvariadic> : PredOpTrait<
+  "the front operand of variadic and non-variadic operand should have the same data type", CPred<[{
+    ([&](const mlir::OperandRange &operands, const mlir::Value &op) {
+      // Retrive all types of operands in the variadic.
+      ValueTypeRange<OperandRange> types = operands.getTypes();
+
+      // If variadic is empty (no front operand), returns true
+      if (types.empty())
+        return true;
+
+      if (mlir::isa<ControlType>(types.front())) {
+        // Pass if non-variadic type is also ControlType
+        return mlir::isa<ControlType>(op.getType());
+      }
+
+      // Pass if non-variadic type has the same data type
+      return mlir::cast<ChannelType>(types.front()).getDataType() == mlir::cast<ChannelType>(op.getType()).getDataType();
+    })
+  }] # "($" # variadic # ", $" # nonvariadic # ")">
+>;
+
+/// Multi-entity constraint ensuring that the all operands of variadic and non-variadic
+/// operand have the same data types.
+/// Also applies to ControlTypes, as some operations treat them interchangeably.
+class AllDataTypesMatchWithVariadic<string variadic, string nonvariadic> : PredOpTrait<
+  "the operands inside the variadic " # name # " and " # nonvariadic # " should all have the same type",
+  And<[
+    // Ensuring the data types of variadic[0], ..., variadic[n] is the same
+    AllDataTypesMatchWithinVariadic<variadic>.predicate,
+    // Ensuring the data types of variadic[0] and nonvariadic is the same
+    DataTypesMatchWithVariadicFront<variadic, nonvariadic>.predicate
+  ]>
+>;
+
+/// Constraint to ensure an input/output is of SimpleControl/Channel type.
+class IsSimpleHandshake<string name> : PredOpTrait<
+  name # " should be of SimpleControl or SimpleChannel type",
+  CPred<"cast<::dynamatic::handshake::ExtraSignalsTypeInterface>($" # name #
+    ".getType()).getNumExtraSignals() == 0">
+>;
+
+/// Merging Relationship of Extra Signals:
+/// the output has the extra signal A if and only if
+/// there exists an input which has the extra signal A.
+/// In Merge-like Ops inputs are always provided as variadic.
+class MergingExtraSignals<string inputs, string output> : PredOpTrait<
+  "the extra signals of intraOperands and interOperands should satisfy \\\"Merging Relationship of Extra Signals\\\"",
```

Comment:
update comments to new names: probably do a ctrl+f to find them all?

Reply:
Do we need a new predicate for select signals being tagless/have you done this somewhere and I just haven't seen it/should it be included in a different predicate?